src/Builder/AbstractFormContractor.php

```diff
@@ -0,0 +1,195 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Builder;
+
+use Sonata\AdminBundle\Admin\AdminInterface;
+use Sonata\AdminBundle\Admin\FieldDescriptionInterface;
+use Sonata\AdminBundle\Form\Type\AdminType;
+use Sonata\AdminBundle\Form\Type\ModelAutocompleteType;
+use Sonata\AdminBundle\Form\Type\ModelHiddenType;
+use Sonata\AdminBundle\Form\Type\ModelListType;
+use Sonata\AdminBundle\Form\Type\ModelReferenceType;
+use Sonata\AdminBundle\Form\Type\ModelType;
+use Sonata\AdminBundle\Form\Type\ModelTypeList;
+use Sonata\Form\Type\CollectionType;
+use Symfony\Component\Form\Extension\Core\Type\FormType;
+use Symfony\Component\Form\FormFactoryInterface;
+
+abstract class AbstractFormContractor implements FormContractorInterface
+{
+    /**
+     * @var FormFactoryInterface
+     */
+    protected $formFactory;
+
+    public function __construct(FormFactoryInterface $formFactory)
+    {
+        $this->formFactory = $formFactory;
+    }
+
+    public function fixFieldDescription(AdminInterface $admin, FieldDescriptionInterface $fieldDescription)
+    {
+        $fieldDescription->setOption('edit', $fieldDescription->getOption('edit', 'standard'));
+
+        // NEXT_MAJOR: Change "$this->hasAssociation($fieldDescription)" with: "$fieldDescription->describesAssociation()".
+        if ($this->hasAssociation($fieldDescription) || $fieldDescription->getOption('admin_code')) {
+            $admin->attachAdminClass($fieldDescription);
+        }
+    }
+
+    /**
+     * @return FormFactoryInterface
+     */
+    public function getFormFactory()
+    {
+        return $this->formFactory;
+    }
+
+    public function getFormBuilder($name, array $formOptions = [])
+    {
+        return $this->getFormFactory()->createNamedBuilder($name, FormType::class, null, $formOptions);
+    }
+
+    public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescription)
+    {
+        // NEXT_MAJOR: Remove this line and update the function signature.
+        $formOptions = \func_get_args()[2] ?? [];
+
+        $options = [];
+        $options['sonata_field_description'] = $fieldDescription;
+
+        if ($this->isAnyInstanceOf($type, [
+            ModelType::class,
+            ModelTypeList::class,
+            ModelListType::class,
+            ModelHiddenType::class,
+            ModelAutocompleteType::class,
+            ModelReferenceType::class,
```

Comment:
This is not present in persistence bundle: https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/src/Builder/FormContractor.php#L111-L115

Reply:
I saw it https://github.com/sonata-project/SonataAdminBundle/pull/3971 in and looking at the implementation:

https://github.com/sonata-project/SonataAdminBundle/blob/6292c61255e945a0aa1440db260680c40de79a14/src/Form/Type/ModelReferenceType.php#L45-L46

it also needs `model_manager` and `class`, never use it though.